[mlir] Add helper to check elementwise-mappable ops with tensors and scalars

snarang181 opened the pull request:
This patch introduces a more general helper for identifying elementwise-mappable operations. The existing utility, `isElementwiseMappableOpOnRankedTensors`, only accepted operations when all operands were ranked tensors. In practice, many elementwise operations in MLIR allow mixing tensor operands with scalars. 
The new helper relaxes the restriction by accepting operands that are either ranked tensors or “scalar-like” types.

snarang181 committed:
Fix TODO to use any_of instead of all_of
Make check more adaptive to include broadcasting of scalars

snarang181 committed:
Add tests

adam-smnk commented on mlir/test/Dialect/Linalg/convert-elementwise-to-linalg.mlir:
```diff
@@ -19,6 +19,53 @@ func.func @addf_rank0(%arg0: tensor<f32>, %arg1: tensor<f32>) -> tensor<f32> {
+// Test a binary elementwise op with a tensor and a scalar operand.
+// CHECK-LABEL: func @addf_tensor_plus_scalar_rank1
+//  CHECK-SAME:   %[[T:[0-9a-zA-Z]*]]: tensor<?xf32>, %[[S:[0-9a-zA-Z]*]]: f32
+func.func @addf_tensor_plus_scalar_rank1(%t: tensor<?xf32>, %s: f32) -> tensor<?xf32> {
```
These tests still don't exercise the new logic. The rewrite occurs on `arith` ops that have tensor only operands.
I'm not sure if there are any upstream ops which are both `ElementwiseMappable` and accept mixed scalar/shape type operands.

adam-smnk replied:
After a bit of digging, I've found there's this test op that should also work with this conversion pass:
```mlir
func.func @test(%arg0: f32, %arg1: tensor<?x?xf32>) {
  %0 = "test.elementwise_mappable"(%arg0, %arg1) : (f32, tensor<?x?xf32>) -> tensor<?x?xf32>
  return
}
```

snarang181 committed:
Classifies scalar-like operands and assigns them a
rank-aware scalar map (d0,…,dn) -> () during lowering.

snarang181 committed:
Fix tests

adam-smnk commented on mlir/test/Dialect/Linalg/convert-elementwise-to-linalg.mlir:
```diff
@@ -108,3 +108,61 @@ func.func @cmpf(%arg0: tensor<4x?x?x8x2x?xf32>, %arg1: tensor<4x?x?x8x2x?xf32>)
 
+// -----
+
+// Check a mix of scalar and tensor input.
```
To be consistent with other test cases, let's interleave checks and ops.

snarang181 committed:
Address review comments

snarang181 commented:
@adam-smnk -- Thank you for your review, I addressed your comments. 
 I still have 
```cpp
 bool anyRankedTensor = llvm::any_of(types, llvm::IsaPred<RankedTensorType>);
```
to meet the rank = 0 operand requirements / test cases.

adam-smnk reviewed (approved):
LGTM

snarang181 commented:
> > to meet the rank = 0 operand requirements / test cases.
> 
> 
> 
> Could you also add a negative test for this?

Sorry, would you mind expanding a bit more?

snarang181 committed:
Add negative test
